### ISSUE TITLE
Clarify if MXIDs are case-sensitive

### ISSUE THREAD

**Issue by ShadowJonathan** [matrix-org/matrix-spec#996]
**Link to problem area**: (the spec in general, or more specifically, the absence of explanation regarding this)

**Issue**
Related to https://github.com/matrix-org/matrix-spec/issues/133

In the spec, there’s no clarification how case-sensitivity interacts with identifying users.

Precedent is that, from implementations, there exist “canonical” casings for MXIDs, but that they’re coalesced in a case-insensitive way from lookups and other addressings.

Conduit has already effectively disabled case-sensitive signups, and converts users on signup to lowercase.

Synapse historically allowed case-sensitive usernames.

The main question here is; if I have a MXID `@Jo:matrix.org` and `@jo:matrix.org`, would these point to the same user, or is it possible these point to two different users?

(I understand the historical nature of this problem, but I think it’s useful to put this problem to bed once and for all by clarifying if MXIDs are case-sensitive or not.)

---

keywords; case insensitive, lowercase, case-sensitive

**Comment by turt2live** [comment]
The spec should already be very clear on this - what do the appendices say?

Users differing only by case are two different users. Synapse collapses this down at registration to try and check for potential logical conflicts, and on login if there's no conflicts then I believe it maps to the correct user ID. It can do this because the username does not need to be the localpart - it just happens to be in many cases.

**Comment by ShadowJonathan** [comment]
> Synapse collapses this down at registration to try and check for potential logical conflicts, and on login if there's no conflicts then I believe it maps to the correct user ID.

Yes, but this is an implementation detail, and the spec does not explicitly spell out that this is correct (in this case).

As an example of where this matters, I was thinking for `ruma` to make its implementation accept case-insensitive `==` for `UserId` (MXID), but I realised the spec does not explicitly spell out that this is fine to do, thus this issue.

The latter cascades into a whole bunch of things, for example one where Conduit might treat `@Jo:matrix.org` as a different user to `@jo:matrix.org`, while synapse would treat them the same.

**Comment by turt2live** [comment]
I'm trying to figure out where we'd put the few sentences that are needed to clarify the situation, though I also believe they're already there in the depths of the spec. Is there a particular area you were looking in?

Element (Web) applies artificial restrictions to ensure lowercase user IDs get created, and Synapse in particular further assists in this by preventing *new* user IDs created by it being upper/mixed case. The login API then tries to map lowercase usernames to mixed/upper case localparts (because in Synapse your username happens to be your localpart - this isn't a requirement in the spec). If Synapse finds that, during login, there are multiple people who could match that username then it requires exact case to be used to continue with login, which is one of the reasons why Element doesn't apply validation to the login page.

**Comment by ShadowJonathan** [comment]
> Element (Web) applies artificial restrictions to ensure lowercase user IDs get created

(note; I wasn't talking about registration, but when inviting people who have an uppercase letter in their name)

> Is there a particular area you were looking in?

Not specifically, though as a pointer I could say that any endpoint - on client-server - which deals with MXIDs, could classify.

Though, now that I'm thinking about this, it's possible that synapse *could* implicitly be doing this on federation endpoints as well (`/invite/')

However, at this point I'm unsure about most things I'm speculating/assuming to be there, this warrants more investigation on my side to figure out what exact behaviour synapse has that fits this description of "upgrading" casings in any join, invite, or user referral (profile requests, etc.) operation.

If I find none, and that element has been correcting this quirk the whole time, then I'll close this issue, otherwise I'll post more specific endpoints and their implementation behaviours.
